ddtrace/profiling/collector/_memalloc_tb.c

```diff
@@ -20,6 +17,77 @@ static PyObject* empty_string = NULL;
 
 static PyObject* ddframe_class = NULL;
 
+#ifndef MEMALLOC_BUFFER_POOL_CAPACITY
+#define MEMALLOC_BUFFER_POOL_CAPACITY 4
+#endif
+
+/* memalloc_tb_buffer_pool is a pool of scratch buffers used for collecting
+ * tracebacks. We don't know ahead of time how many frames a traceback will have,
+ * and traversing the list of frames can be expensive. At the same time, we want
+ * to right-size the tracebacks we aggregate in memory to minimize waste. So we
+ * use scratch buffers of the maximum configured traceback size and then copy
+ * the actual traceback once we know how many frames it has.
+ *
+ * We need a pool because, for some Python versions, collecting a traceback
+ * releases the GIL. So we can't just have one scratch buffer or we will have a
+ * logical race in writing to the buffer. The calling thread owns the scratch
+ * buffer it gets from the pool until the buffer is returned to the pool.
+ */
+typedef struct
+{
+    /* TODO: if/when we support no-GIL or subinterpreter python, we'll need a
+     * lock to protect the pool in get & put */
+    traceback_t* pool[MEMALLOC_BUFFER_POOL_CAPACITY];
+    size_t count;
+    size_t capacity;
+} memalloc_tb_buffer_pool;
+
+/* For now we use a global pool */
+static memalloc_tb_buffer_pool g_memalloc_tb_buffer_pool = {
+    .count = 0,
+    .capacity = MEMALLOC_BUFFER_POOL_CAPACITY,
+};
+
+#undef MEMALLOC_BUFFER_POOL_CAPACITY
+
+static traceback_t*
+memalloc_tb_buffer_pool_get(memalloc_tb_buffer_pool* pool, uint16_t max_nframe)
```

Comment:
We just simply assume that the GIL will be held when this function is called, right? Do we check from the functions that call into this?